opers/kafka_in.go

```diff
@@ -77,14 +81,53 @@ func (k *KafkaInOperator) IngestBatch(recordBatchBytes []byte, processor proc.Pr
 	processor.GetReplicator().ReplicateBatch(processBatch, complFunc)
 }
 
+func (k *KafkaInOperator) maybeHandleIdempotentProducerBatch(partitionID, processorID int, bytes []byte) error {
+	producerID := int(binary.BigEndian.Uint64(bytes[43:51]))
+	// if the producer doesn't have idempotency enabled, the producer id will be -1
+	if producerID > -1 {
+		baseSequence := int(binary.BigEndian.Uint32(bytes[53:57]))
+		lastOffsetDelta := int(binary.BigEndian.Uint32(bytes[23:27]))
+		sequenceNumber := baseSequence + lastOffsetDelta
+		if processorID >= 0 && processorID < len(k.partitionProducerMapping) {
+			if k.partitionProducerMapping[processorID] == nil {
+				k.partitionProducerMapping[processorID] = make(map[int]map[int]int)
+			}
+		} else {
+			panic("unexpected processor ID")
+		}
+		if k.partitionProducerMapping[processorID][partitionID] == nil {
```

Comment:
If you store this in a local variable you can avoid the double lookup multiple times